src/titan_checkpoint_impl.cc

```diff
@@ -0,0 +1,392 @@
+                return CopyFile(db_->GetEnv(), src_dirname + fname,
+                                full_private_path + fname, size_limit_bytes,
+                                db_options.use_fsync);
+              } /* copy_file_cb */,
+              [&](const std::string& fname, const std::string& contents, FileType) {
+                ROCKS_LOG_INFO(db_options.info_log, "Creating %s", fname.c_str());
+                return CreateFile(db_->GetEnv(), full_private_path + fname, contents,
+                                  db_options.use_fsync);
+              } /* create_file_cb */,
+              &sequence_number, log_size_for_flush, full_private_path);
+    }
+
+    if (disabled_file_deletions) {
+      // We copied all the files, enable file deletions
+      Status ss = db_->EnableFileDeletions(false);
+      assert(ss.ok());
+    }
+
+  }
+
+  if (s.ok()) {
+    // Move tmp private backup to real checkpoint directory
+    s = db_->GetEnv()->RenameFile(full_private_path, checkpoint_dir);
+  }
+  if (s.ok()) {
+    std::unique_ptr<Directory> checkpoint_directory;
+    db_->GetEnv()->NewDirectory(checkpoint_dir, &checkpoint_directory);
+    if (checkpoint_directory != nullptr) {
+      s = checkpoint_directory->Fsync();
+    }
+  }
+
+  if (s.ok()) {
+    // Here we know that we succeeded and installed the new checkpoint
+    ROCKS_LOG_INFO(db_options.info_log, "Checkpoint DONE. All is good");
+    ROCKS_LOG_INFO(db_options.info_log, "Checkpoint sequence number: %" PRIu64,
+                   sequence_number);
+  } else {
+    // Clean all the files we might have created
+    ROCKS_LOG_INFO(db_options.info_log, "Checkpoint failed -- %s",
+                   s.ToString().c_str());
+    CleanStagingDirectory(full_private_path, db_options.info_log.get());
+  }
+  return s;
+}
+
+Status TitanCheckpointImpl::CreateCustomCheckpoint(
+        const DBOptions& db_options,
+        std::function<Status(const std::string& src_dirname,
+                             const std::string& src_fname, FileType type)>
+        link_file_cb,
+        std::function<Status(const std::string& src_dirname,
+                             const std::string& src_fname,
+                             uint64_t size_limit_bytes, FileType type)>
+        copy_file_cb,
+        std::function<Status(const std::string& fname, const std::string& contents,
+                             FileType type)>
+        create_file_cb,
+        uint64_t* sequence_number, uint64_t log_size_for_flush,
+        const std::string full_private_path) {
+  Status s;
+  std::vector<std::string> live_files;
+  std::vector<std::string> titan_live_files;
+  std::vector<VersionEdit> version_edits;
+  uint64_t base_manifest_file_size = 0;
+  uint64_t min_log_num = port::kMaxUint64;
+  *sequence_number = db_->GetLatestSequenceNumber();
+  bool same_fs = true;
+  VectorLogPtr live_wal_files;
+
+  bool flush_memtable = true;
+  if (!db_options.allow_2pc) {
```

Comment:
Can we make the process like that:
1. DisableFileDeletions, this will prevent blob files being deleted
2. Call db_->CreateCustomCheckpoint without mutex to get a checkpoint of base db
3. Collect all file（live + obsolete）about titandb and record all blob files into version_edit(prevent removing useful blob file when Titan reopen)
4. EnableFileDeletions

If that, we can reuse the logic of `CreateCustomCheckpoint` of RocksDB

Reply:
I reused the the logic of `CreateCustomCheckpoint` in new commit, and modified `CurrentFileModifiedWhileCheckpointing` test case.